### PR TITLE
Np 3592 get organization without hierarchy

### PR DESCRIPTION
**Pull request by svenngjeto**

### CODE REVIEW

File: cristin-commons/src/main/java/no/unit/nva/utils/UriUtils.java [hunk 1]
```diff
@@ -84,6 +84,8 @@ public static URI getCristinUri(String identifier, String path) {
                 .getUri();
     }
 
+
```
Comment (torarnet) [inline cristin-commons/src/main/java/no/unit/nva/utils/UriUtils.java:87]: Remove whitespace?

File: cristin-organization/src/test/java/no/unit/nva/cristin/organization/organization.feature [hunk 2]
```diff
@@ -59,3 +60,19 @@ Feature: API tests for Cristin Organization retrieve and search
...
+    And match response.hits == '#array'
+
+  Scenario: GET organization for known organization returns list of search results without depth
+    Given  path '/organization'
+    And param query = existingOrganizationIdentifier
+    And param depth = 'top'
+    When method GET
+    Then status 200
+    And match response.hits == '#array'
+    And match response.hits[0].partOf != '#present'
```
Comment (torarnet) [inline cristin-organization/src/test/java/no/unit/nva/cristin/organization/organization.feature:78]: So we dont want upper organizations this hit if part of? Is not that what top is? get all organizations this is part of? And rather exclude hasPart?

File: cristin-organization/src/test/java/no/unit/nva/cristin/organization/organization.feature [hunk 3]
```diff
@@ -59,3 +60,19 @@ Feature: API tests for Cristin Organization retrieve and search
     And match response.hits == '#array'
     And match response.size == '#number'
     And match response.hits == '#[2]' // hits array length == 0
+
+  Scenario: GET organization for known organization returns list of search results with depth
+    Given  path '/organization'
+    And param query = existingOrganizationIdentifier
+    When method GET
+    Then status 200
+    And match response.hits == '#array'
```
Comment (torarnet) [inline cristin-organization/src/test/java/no/unit/nva/cristin/organization/organization.feature:69]: Should we add hasPart == present?

File: cristin-organization/src/test/java/no/unit/nva/cristin/organization/organization.feature (hunk 2 above)
Comment (torarnet) [reply cristin-organization/src/test/java/no/unit/nva/cristin/organization/organization.feature:78]: Or is parameter top to get top level organization for given identifier?

File: cristin-organization/src/main/java/no/unit/nva/cristin/organization/CristinOrganizationApiClient.java [hunk 4]
```diff
@@ -80,6 +90,41 @@ public Organization getOrganization(URI uri)
...
 
+    /**
+     * Get information for an Organization without parent of subunits.
+     *
+     * @param identifier the Cristin unit URI
+     * @return an {@link Organization} containing the information
+     * @throws NotFoundException when the URI does not correspond to an existing unit.
+     */
+    public Organization getFlatOrganization(String identifier) throws ApiGatewayException, InterruptedException {
+        URI cristinUri = new UriWrapper(CRISTIN_API_URL)
```
Comment (torarnet) [inline cristin-organization/src/main/java/no/unit/nva/cristin/organization/CristinOrganizationApiClient.java:101]: extract method?

File: cristin-organization/src/test/java/no/unit/nva/cristin/organization/FetchCristinOrganizationHandlerTest.java [hunk 5]
```diff
@@ -210,6 +212,44 @@ void shouldReturnHttp404NotFoundWhenNonExistingIdentifier() throws Exception {
...
+        final URI level2a = getCristinUri("185.53.0.0", UNITS_PATH);
+        doReturn(getSubSubUnit("unit_18_53_0_0.json")).when(mySpy).getSubSubUnitDtoWithMultipleEfforts(level2a);
+        final URI level2b = getCristinUri("185.50.0.0", UNITS_PATH);
+        doReturn(getSubSubUnit("unit_18_50_0_0.json")).when(mySpy).getSubSubUnitDtoWithMultipleEfforts(level2b);
+        final URI level3 = getCristinUri("185.53.18.0", UNITS_PATH);
+        doReturn(getSubSubUnit("unit_18_53_18_0.json")).when(mySpy).getSubSubUnitDtoWithMultipleEfforts(level3);
+        final URI level4 = getCristinUri("185.53.18.14", UNITS_PATH);
+        doReturn(getSubSubUnit("unit_18_53_18_14.json")).when(mySpy).getSubSubUnitDtoWithMultipleEfforts(level4);
+
+        fetchCristinOrganizationHandler = new FetchCristinOrganizationHandler(cristinApiClient, new Environment());
```
Comment (torarnet) [inline cristin-organization/src/test/java/no/unit/nva/cristin/organization/FetchCristinOrganizationHandlerTest.java:235]: You are doing this two different places in the same method

File: cristin-organization/src/test/java/no/unit/nva/cristin/organization/FetchCristinOrganizationHandlerTest.java [hunk 6]
```diff
@@ -210,6 +212,44 @@ void shouldReturnHttp404NotFoundWhenNonExistingIdentifier() throws Exception {
     }
 
 
+    @Test
+    void shouldReturnOrganizationFlatHierarchy() throws IOException, ApiGatewayException {
+
+        HttpClient httpClient = mock(HttpClient.class);
+        when(httpClient.sendAsync(any(), any())).thenThrow(new RuntimeException("This should Not happen!"));
+        output = new ByteArrayOutputStream();
+        fetchCristinOrganizationHandler = new FetchCristinOrganizationHandler(cristinApiClient, new Environment());
```
Comment (torarnet) [inline cristin-organization/src/test/java/no/unit/nva/cristin/organization/FetchCristinOrganizationHandlerTest.java:221]: See other comment

File: cristin-organization/src/test/java/no/unit/nva/cristin/organization/FetchCristinOrganizationHandlerTest.java [hunk 7]
```diff
@@ -210,6 +212,44 @@ void shouldReturnHttp404NotFoundWhenNonExistingIdentifier() throws Exception {
...
+        doReturn(getSubSubUnit("unit_18_53_0_0.json")).when(mySpy).getSubSubUnitDtoWithMultipleEfforts(level2a);
+        final URI level2b = getCristinUri("185.50.0.0", UNITS_PATH);
+        doReturn(getSubSubUnit("unit_18_50_0_0.json")).when(mySpy).getSubSubUnitDtoWithMultipleEfforts(level2b);
+        final URI level3 = getCristinUri("185.53.18.0", UNITS_PATH);
+        doReturn(getSubSubUnit("unit_18_53_18_0.json")).when(mySpy).getSubSubUnitDtoWithMultipleEfforts(level3);
+        final URI level4 = getCristinUri("185.53.18.14", UNITS_PATH);
+        doReturn(getSubSubUnit("unit_18_53_18_14.json")).when(mySpy).getSubSubUnitDtoWithMultipleEfforts(level4);
+
+        fetchCristinOrganizationHandler = new FetchCristinOrganizationHandler(cristinApiClient, new Environment());
+        final String identifier = "185.53.18.14";
```
Comment (torarnet) [inline cristin-organization/src/test/java/no/unit/nva/cristin/organization/FetchCristinOrganizationHandlerTest.java:236]: Constant?

File: cristin-organization/src/test/java/no/unit/nva/cristin/organization/organization.feature [hunk 8]
```diff
@@ -59,3 +60,19 @@ Feature: API tests for Cristin Organization retrieve and search
     And match response.hits == '#array'
     And match response.size == '#number'
     And match response.hits == '#[2]' // hits array length == 0
+
```
Comment (torarnet) [inline cristin-organization/src/test/java/no/unit/nva/cristin/organization/organization.feature:63]: Should you add a test with depth = none?

File: cristin-organization/src/test/java/no/unit/nva/cristin/organization/FetchCristinOrganizationHandlerTest.java [hunk 9]
```diff
@@ -210,6 +212,44 @@ void shouldReturnHttp404NotFoundWhenNonExistingIdentifier() throws Exception {
...
+        GatewayResponse<Organization> gatewayResponse = GatewayResponse.fromOutputStream(output);
+
+        assertEquals(HTTP_OK, gatewayResponse.getStatusCode());
+        assertThat(gatewayResponse.getHeaders(), hasKey(CONTENT_TYPE));
+        assertThat(gatewayResponse.getHeaders(), hasKey(ACCESS_CONTROL_ALLOW_ORIGIN));
+
+        URI expectedId = getNvaApiId(identifier, ORGANIZATION_PATH);
+        Organization actualOrganization = gatewayResponse.getBodyObject(Organization.class);
+        assertEquals(actualOrganization.getId(), expectedId);
+        assertThat(actualOrganization.getName().get("en"), containsString("Department of Medical Biochemistry"));
```
Comment (torarnet) [inline cristin-organization/src/test/java/no/unit/nva/cristin/organization/FetchCristinOrganizationHandlerTest.java:247]: Also check that hasPart and partOf is not present?

File: cristin-organization/src/main/java/no/unit/nva/cristin/organization/CristinOrganizationApiClient.java [hunk 10]
```diff
@@ -80,6 +90,41 @@ public Organization getOrganization(URI uri)
...
+                .addChild(UNITS_PATH)
+                .addQueryParameter(UNIT_ID, identifier)
+                .getUri();
+
+        HttpResponse<String> response = sendRequestMultipleTimes(addLanguage(cristinUri)).get();
+        if (isSuccessful(response.statusCode())) {
+            try {
```
Comment (torarnet) [inline cristin-organization/src/main/java/no/unit/nva/cristin/organization/CristinOrganizationApiClient.java:108]: Do all these lines need to be in the try catch? Or can you use attempt() orElseThrow?

File: cristin-organization/src/main/java/no/unit/nva/cristin/organization/CristinOrganizationApiClient.java [hunk 11]
```diff
@@ -80,6 +90,41 @@ public Organization getOrganization(URI uri)
...
+                List<SubUnitDto> units = OBJECT_MAPPER.readValue(response.body(), new TypeReference<>() {
+                });
+                if (SINGLE_HIT == units.size()) {
+                    SubUnitDto subUnitDto = units.get(FIRST_AND_ONLY_UNIT);
+                    return new Organization.Builder()
```
Comment (torarnet) [inline cristin-organization/src/main/java/no/unit/nva/cristin/organization/CristinOrganizationApiClient.java:113]: extract method or keep as is? Thinking for readability

**Review by torarnet — COMMENTED** [review]
See comments

**Review by brinxmat — COMMENTED** [review]
See other reviewer's comments

File: cristin-organization/build.gradle [hunk 12]
```diff
@@ -9,6 +9,18 @@ dependencies {
     testImplementation libs.karate.junit5
 }
 
+tasks.named('test') {
```
[the remark this answers is not in the document]
Comment (svenngjeto) [reply cristin-organization/build.gradle:12]: Added to exclude karatetests from regular build

File: cristin-organization/src/main/java/no/unit/nva/cristin/organization/CristinOrganizationApiClient.java (hunk 4 above)
Comment (svenngjeto) [reply cristin-organization/src/main/java/no/unit/nva/cristin/organization/CristinOrganizationApiClient.java:126]: Extracted some code

File: cristin-organization/src/test/java/no/unit/nva/cristin/organization/FetchCristinOrganizationHandlerTest.java (hunk 5 above)
Comment (svenngjeto) [reply cristin-organization/src/test/java/no/unit/nva/cristin/organization/FetchCristinOrganizationHandlerTest.java:245]: It's duplicated to have the same base return different with regards to query parameters

File: cristin-organization/src/test/java/no/unit/nva/cristin/organization/organization.feature (hunk 8 above)
Comment (svenngjeto) [reply cristin-organization/src/test/java/no/unit/nva/cristin/organization/organization.feature:64]: yes, done - as i should be in the first place

**Review by brinxmat — DISMISSED** [review]

File: cristin-organization/src/main/java/no/unit/nva/cristin/organization/CristinOrganizationApiClient.java [hunk 13]
```diff
@@ -79,6 +89,46 @@ public Organization getOrganization(URI uri) throws ApiGatewayException  {
...
+     * @return an {@link Organization} containing the information
+     */
+    public Organization getFlatOrganization(String identifier) {
+        URI cristinUri = getCristinOrganizationByIdentifierUri(identifier);
+        HttpResponse<String> response = sendRequestMultipleTimes(addLanguage(cristinUri)).get();
+        return isSuccessful(response.statusCode()) ? extractOrganization(identifier, response) : null;
+    }
+
+    private Organization extractOrganization(String identifier, HttpResponse<String> response) {
+        return attempt(() -> {
```
Comment (torarnet) [inline cristin-organization/src/main/java/no/unit/nva/cristin/organization/CristinOrganizationApiClient.java:105]:
```suggestion
 List<SubUnitDto> units = attempt(() -> { OBJECT_MAPPER.readValue(response.body(), new TypeReference<>(){}) }).orElseThrow(fail -> JsonException...);
 if (SINGLE_HIT == units.size()) {
   return toOrganization(identifier, units.get(FIRST_AND_ONLY_UNIT));
 } else {
   throw new BadRequestException(UNIQUELY_IDENTIFY_ORGANIZATION);
 }
```

**Review by torarnet — DISMISSED** [review]
